pythran/toolchain.py

```diff
@@ -299,31 +228,44 @@ def compile_cxxfile(cxxfile, module_so=None, **kwargs):
     Raises CompileError on failure
 
     '''
-    # FIXME: not sure about overriding the user defined compiler here...
-    compiler = kwargs.get('cxx', default_compiler())
-
-    _cppflags = cppflags() + kwargs.get('cppflags', [])
-    _cxxflags = cxxflags() + kwargs.get('cxxflags', [])
-    _ldflags = ldflags() + kwargs.get('ldflags', [])
-
-    # Get output filename from input filename if not set
-    module_so = module_so or (os.path.splitext(cxxfile)[0] + ".so")
-    try:
-        cmd = ([compiler, cxxfile] +
-               _cppflags +
-               _cxxflags +
-               ["-shared", "-o", module_so] +
-               _ldflags)
-        logger.info("Command line: " + _format_cmdline(cmd))
-        output = check_output(cmd, stderr=STDOUT)
-    except CalledProcessError as e:
-        raise CompileError(e.cmd, e.output)
-    except Exception:
-        print("E: Error encountered while running:")
-        print(''.join(cmd))
-        raise
-    logger.info("Generated module: " + module_so)
-    logger.info("Output: " + output)
+    if module_so:
+        module_name, _ = os.path.splitext(os.path.basename(module_so))
+    else:
+        module_name, _ = os.path.splitext(os.path.basename(cxxfile))
+
+    builddir = mkdtemp()
+    buildtmp = mkdtemp()
+
+    extension_args = make_extension(**kwargs)
+
+    extension = Extension(module_name,
+                          [cxxfile],
+                          language="c++",
+                          **extension_args)
+
+    setup(name=module_name,
+          ext_modules=[extension],
+          # fake CLI call
+          script_name='setup.py',
+          script_args=['--verbose'
+                       if logger.isEnabledFor(logging.INFO)
+                       else '--quiet',
+                       'build_ext',
+                       '--build-lib', builddir,
+                       '--build-temp', buildtmp,
+                       ]
+          )
```

Comment:
> I think it would be better to keep raison CompilationError in case of error here as it is more explicite than SystemExit.
you mean catching SytemError and raising CompilationError?